docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,196 @@
+| R2 | The machine health check controller MUST be able to select machines by label which belong to a MachinePool and remediate. | U2 |
+| R3 | The MachinePool API MUST provide an optional deployment strategy using the same type as MachineDeployment.Spec.Strategy. | U3 |
+| R4 | The Machine Controller MUST handle node drain for Machine Pool Machines with the same behavior as MachineDeployment Machines. | U1, U3 |
+
+### Implementation Details/Notes/Constraints
+
+- As an alternative, the `infrastructureRefList` field could be attached to the provider-specific MachinePool resource. Then the CAPI MachinePool controller could check through the provider-specific MachinePool resource through its infraRef to determine if it has an `infrastructureRefList` and use it if so. This would be very similar to the current design, except it would avoid changing the MachinePool spec in CAPI.
+- The `infrastructureRefList` might be more appropriate on MachinePool.Status instead of MachinePool.Spec.
+- It is possible that the `infrastructureRefList` could become large, if there are many machines in the MachinePool. It's a small struct, but carrying a long list around isn't optimal. One alternative suggested is to use a Kubernetes Selector similarly to MachineDeployment, but it's unclear how it would be used since the code in question is creating Machines, not finding existing ones.
+
+### Risks and Mitigations
+
+- It is conceivable that the `infrastructureRefList` could contain thousands of instances and that the representation of the MachinePool could cause storage or network transfer issues. As mentioned above, alternatives to this field are welcome for consideration.
+
+## Alternatives
+
+Refactor MachineDeployment/MachineSet to support the option of using cloud infrastructure-native scaling group resources. That is, merge MachinePool into MachineDeployment.
+
+MachinePool and MachineDeployment overlap significantly, and this proposal aims to eliminate one of their major differences: that individual Machines aren’t represented in a MachinePool. While it is valuable that MachineDeployments are provider-agnostic, MachinePools take advantage of scalable resources unique to an infra provider, which may have advantages in speed or reliability. To merge MachinePool behavior — along with the changes proposed here — into MachineDeployment would effectively deprecate the experimental MachinePool API and would increase the overall scope of work significantly, while increasing challenges to maintaining backward compatibility in the stable API.
+
+<!-- TODO: add other pro + con arguments? This comes up a lot. -->
+
+## Upgrade Strategy
+
+MachinePool Machines will be an optional, backward-compatible feature. Existing infrastructure providers can make no changes and will observe the same behavior with MachinePools they always have.
```

Comment:
Is there any reason to preserve both scenarios long term? Otherwise we might want to have a deprecation period for providers to adapt and so eventually only MachinePool with Machines is supported. Otherwise we just have an additional scenario to maintain.

Reply:
That's a good question, I'm not too sure. Any thoughts @mboersma @devigned?

Reply:
I think it's possible that a cloud provider has a native scalable resource (like an Amazon ASG or Azure VMSS) that doesn't support removing individual instances. In that case, the behavior should stay as it is (no MachinePool Machines created).

Or we could do what @fabriziopandini hinted at above: create MachinePool Machines that either can't be deleted or maybe log a warning on deletion that points out no cloud resources were deleted. That would make for a more consistent UX, but arguably it's another dual behavior scenario.

It's also possible that my hypothetical--a scalable resource that doesn't allow deleting individual instances--doesn't actually exist, and shouldn't be factored in to this. I'm not familiar with all the CAPI providers...

Reply:
I think my hypothetical – a scalable resource that doesn’t allow deleting individual instances – is a red herring. It either doesn’t exist or doesn’t have a MachinePool implementation, as best I can tell from looking at CAPI providers.

So I’m tending towards @enxebre’s suggestion of having a deprecation/implementation period.

Reply:
Since we learned that there are indeed MachinePool implementations that don't allow deleting individual instances, this proposal has to allow for the variation in UX and keep this as an optional feature.